src/odemis/driver/piezomotor.py

```diff
@@ -0,0 +1,733 @@
+            except KeyError:
+                closed_loop = True
+                logging.info("Axis mode (closed/open loop) not specified for axis %s. Assuming closed loop.", axis_name)
+            self._closed_loop[axis_name] = closed_loop
+
+            ad = model.Axis(canAbs=closed_loop, unit=axis_unit, range=axis_range)
+            axes_def[axis_name] = ad
+
+        Actuator.__init__(self, name, role, axes=axes, inverted=inverted, **kwargs)
+        self._executor = CancellableThreadPoolExecutor(max_workers=1)  # one task at a time
+        self._ser_access = threading.RLock()
+
+        # Connect to hardware
+        self._port = None  # port number
+        min_axis = min(self._axis_map.values())
+        self._serial = self._findDevice(port, min_axis)
+        self._recovering = False
+
+        # Get version
+        hwVersions = []
+        for ax_name, ax_num in self._axis_map.items():
+            ver = self.getVersion(ax_num)
+            sn = self.getSerialNumber(ax_num)
+            hwVersions.append("Axis %s ('%s') version: %s, " % (ax_num, ax_name, ver) +
+                              "serial number: %s" % sn)
+        self._hwVersion = ", ".join(hwVersions)
+
+        # Position and referenced VAs
+        self.position = model.VigilantAttribute({}, unit="m", readonly=True)
+        self.referenced = model.VigilantAttribute({}, unit="m", readonly=True)
+
+        # Configuration
+        for axis in self._axis_map.values():
+            self.setWaveform(axis, WAVEFORM_DELTA)
+
+        driver_name = getSerialDriver(self._port)
+        self._swVersion = "serial driver: %s" % (driver_name,)
+
+    def terminate(self):
+        # terminate can be called several times, do nothing if ._serial is already None
+        if self._serial is None:
+            return
+        self._serial = None
+        for axis in self._axis_map.values():
+            self.setWaveform(axis, WAVEFORM_PARK)  # power off
+        self._serial.close()
+
+    def stop(self, axes=None):
+        self._executor.cancel()
+        axes = axes or self.axes
+        for ax in axes:
+            self.stopAxis(ax)
+
+    def moveRel(self, shift):
+        if not shift:
+            return model.InstantaneousFuture()
+        self._checkMoveRel(shift)
+        shift = self._applyInversion(shift)
+        f = self._executor.submit(self._doMoveRel, shift)
+        return f
+
+    def moveAbs(self, pos):
+        self._checkMoveAbs(pos)
+        pos = self._applyInversion(pos)
+        f = self._executor.submit(self._doMoveAbs, pos)
+        return f
+
+    def reference(self, axes):
+        f = self._executor.submit(self._doReference, axes)
+        return f
+
+    def _doReference(self, axes):
+        self._check_hw_error()
+        for axname in axes:
+            self.startIndexMode(self._axis_map[axname])
+            self.moveToIndex(self._axis_map[axname])
+            if 'index' not in self.getIndexStatus(self._axis_map[axname]):
```

Comment:
You'll probably need to wait (with a timeout) here, as moveToIndex is not blocking. Typically a reference procedure can takes several seconds.